pkg/networkdriver/sriov/utils.go

```
@@ -0,0 +1,172 @@
+	"strings"
+
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+const (
+	defaultSysPath = "/sys/bus/pci/devices"
+	defaultNetPath = "/sys/class/net"
+	// https://elixir.bootlin.com/linux/v6.17.6/source/include/linux/pci_ids.h#L32
+	// #define PCI_CLASS_NETWORK_ETHERNET	0x0200
+	ethernetDeviceClass = 0x0200
+)
+
+// isNetworkDevice checks the PCI device class and returns whether it is a network device.
+// https://elixir.bootlin.com/linux/v6.17.6/source/include/linux/pci_ids.h#L32
+// #define PCI_CLASS_NETWORK_ETHERNET	0x0200
+func isNetworkDevice(pciAddr string) bool {
+	deviceClassPath := path.Join(defaultSysPath, pciAddr, "class")
+	f, err := os.ReadFile(deviceClassPath)
+	if err != nil {
+		return false
+	}
+
+	v, err := strconv.ParseInt(strings.ReplaceAll(string(f), "\n", ""), 0, 32)
+	if err != nil {
+		return false
+	}
+
+	// get just the first 2 bytes, ignore subclass part
+	v = v >> 8
+
+	return v == ethernetDeviceClass
+}
+
+// parseDevice constructs a PciDevice from a PCI device's sysfs attributes.
+func (mgr *SRIOVManager) parseDevice(addr string) *PciDevice {
+	dev := PciDevice{
+		addr: addr,
+	}
+
+	devicePath := path.Join(mgr.sysPath, addr)
+	driver, err := os.Readlink(path.Join(devicePath, "driver"))
+	if err == nil {
+		// 	/sys/bus/pci/devices/0000:02:00.0# readlink driver
+		// ../../../../bus/pci/drivers/mlx5_core
+		_, driver := path.Split(driver)
+		dev.driver = driver
+	}
+
+	vendor, err := os.ReadFile(path.Join(devicePath, "vendor"))
+	if err == nil {
+		dev.vendor = strings.ReplaceAll(string(vendor), "\n", "")
+	}
+
+	device, err := os.ReadFile(path.Join(devicePath, "device"))
+	if err == nil {
+		dev.deviceID = strings.ReplaceAll(string(device), "\n", "")
+	}
+
+	ifNames, err := os.ReadDir(path.Join(devicePath, "net"))
+	if err == nil {
+		for _, d := range ifNames {
+			dev.ifName = d.Name()
+			break
+		}
+	}
+
+	// absurdly nested for now x.x
+	pfNames, err := os.ReadDir(path.Join(devicePath, "physfn", "net"))
+	if err == nil {
```

We should avoid swallowing the errors and propagate them to `ListDevices` instead. `ListDevices` already return an error, so we could simply accumulate all the errors we get here without early-returning and join them at the end of the function. Same in `ListDevices` to propagate them back to `getDevicePools`

---

still not addressed, will revisit this later today

---

i have taken a quick look at it, and the error handling here is pretty bad, and we should fix it.
i created a separate issue to track that one, but for now i'd say lets go ahead with merging this in so we can unblock other workstreams

---

Added the error handling in https://github.com/cilium/cilium/pull/43007/commits/a207ed81814397c908be1c4e143029a4d5a04feb

Let's revisit this later in case we want to define a set of non-mandatory fields.